

1. In this part, you are required to implement Apriori and FPgrowth algorithms on the data provided as "data_apriori_fpgrowth/data.csv". The dataset is of online directory of certified businesses.

2. We have loaded the dataset for you.

3. You are free to use any library or implement your own algorithms.


**Use minimum support : 0.1**

In [0]:
!pip install mlxtend --upgrade --no-deps

In [0]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, fpgrowth
from mlxtend.preprocessing import TransactionEncoder

In [0]:
from google.colab import drive
drive.mount('/content/drive')

In [0]:
!unzip -uq "/content/drive/My Drive/DataMining_hw2/hw2.zip" -d "./"

In [0]:
data = np.load("data_apriori_fpgrowth/data.npy", allow_pickle = True)

In [0]:
minsup=0.1

In [0]:
# preprocessing
te = TransactionEncoder()
te_ary = te.fit(data).transform(data)
df = pd.DataFrame(te_ary, columns=te.columns_)

# To Do
1. Run Apriori Algorithm on the dataset using minimum support 0.1.
 1. Note the run time of the algorithm to find frequent item sets.
 1. Print all the frequent item sets found by apriori.
 1. Print all frequent item sets of length $2$ or higher with support of $0.2$ or higher. 
1. Run FPGrowth Algorithm on the dataset using minimum support 0.1.
 1. Note the run time of the FPGrowth algorithm to find frequent item sets.
 1. Print all frequent item sets found by FPGrowth.
 1. Print all frequent item sets of length $2$ or higher with support of $0.2$ or higher.
1. Answer the following question:
 1. Both algorithms find and return exactly the same frequent itemsets (given same parameters of support and itemset length). What difference do you note in both of the algorithms?
 1. There were more than $1000$ transactions. Why there are too few itemsets returned by the algorithms?
 

# Apriori [10 marks]

In [0]:
# Apriori
%timeit -n 100 -r 10 frequent_itemsets = apriori(df, min_support=minsup, use_colnames=True)

100 loops, best of 10: 9.27 ms per loop


In [0]:
# All Frequent Itemsets
frequent_itemsets

support                       itemsets
0   0.202113                        (ASIAN)
1   0.300704                        (BLACK)
2   0.152113                     (Brooklyn)
3   0.164085                     (HISPANIC)
4   0.671127                          (MBE)
5   0.300000                 (NON-MINORITY)
6   0.295070                     (New York)
7   0.477465                          (WBE)
8   0.200000                   (ASIAN, MBE)
9   0.300704                   (BLACK, MBE)
10  0.112676                (MBE, Brooklyn)
11  0.164085                (MBE, HISPANIC)
12  0.170423                (MBE, New York)
13  0.169014                     (WBE, MBE)
14  0.118310       (NON-MINORITY, New York)
15  0.300000            (NON-MINORITY, WBE)
16  0.175352                (WBE, New York)
17  0.118310  (NON-MINORITY, WBE, New York)

In [0]:
# Frequent itemsets of length 2 with support of 0.2
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets[ (frequent_itemsets['length'] >= 2) & (frequent_itemsets['support'] >= 0.2) ]

support             itemsets  length
8   0.200000         (ASIAN, MBE)       2
9   0.300704         (BLACK, MBE)       2
15  0.300000  (NON-MINORITY, WBE)       2

# FPGrowth [10 marks]

In [0]:
# FPGrowth
%timeit -n 100 -r 10 fp_itemsets = fpgrowth(df, min_support=minsup, use_colnames=True)

100 loops, best of 10: 15 ms per loop


In [0]:
fp_itemsets

support                       itemsets
0   0.671127                          (MBE)
1   0.477465                          (WBE)
2   0.300704                        (BLACK)
3   0.152113                     (Brooklyn)
4   0.202113                        (ASIAN)
5   0.295070                     (New York)
6   0.164085                     (HISPANIC)
7   0.300000                 (NON-MINORITY)
8   0.169014                     (WBE, MBE)
9   0.300704                   (BLACK, MBE)
10  0.112676                (MBE, Brooklyn)
11  0.200000                   (ASIAN, MBE)
12  0.175352                (WBE, New York)
13  0.170423                (MBE, New York)
14  0.118310       (NON-MINORITY, New York)
15  0.118310  (NON-MINORITY, WBE, New York)
16  0.164085                (MBE, HISPANIC)
17  0.300000            (NON-MINORITY, WBE)

In [0]:
# Frequent itemsets of length 2 with support of 0.2
fp_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
fp_itemsets[ (frequent_itemsets['length'] >= 2) & (frequent_itemsets['support'] >= 0.2) ]

support                       itemsets  length
8   0.169014                     (WBE, MBE)       2
9   0.300704                   (BLACK, MBE)       2
15  0.118310  (NON-MINORITY, WBE, New York)       2

# Answer to the Questions [5 marks]

Answer 1: 

* Even though theoreticall the FP-growth algorithm should perform magnitutdes faster than the apriori algorithm, the results show that the FP-growth algorithm is performing slightly slower than it.
* The itemsets of the apriori algorithm are sorted from 1 item, to 2 items, to 3 items, which represents how the apriori algorithm builds it's itemset.

In [0]:
unique = set(x for l in data for x in l)

sum = 0
for i in range(len(data)):
    sum += len(data[i])

mean_len = sum//len(data)
print("Number of Transactions: " + str(len(data)))
print("Average number of items in a transaction: " +str(mean_len))
print("Distinct items in the dataset: "+ str(len(unique)))
print("Required Transaction presence to be frequent: " +str(minsup*len(data)))


Number of Transactions: 1420
Average number of items in a transaction: 4
Distinct items in the dataset: 721
Required Transaction presence to be frequent: 142.0


Answer 2: There are over 700 unique items in the data set with 1400 transactions having took place. The average number of items in a transaction was 4 and one item needed to appear in atleast 142 transactions in order to be frequent. 

With this data, we can see that the best case scenario for the number of frequent itemsets would be ~40 itemsets. which is still very less alongside our ~18 itemsets.

The reason for this is due to the small number of items in a transaction and some items appearing much much more frequently than others (e.g MBE). If there were more items in a transaction then we would see a rise in the number of frequent itemsets